app/code/core/Mage/Checkout/controllers/CartController.php

```diff
@@ -620,6 +620,7 @@ public function couponPostAction()
                 }
             } else {
                 $this->_getSession()->addSuccess($this->__('Coupon code was canceled.'));
+                $this->_getSession()->setCartCouponCode($couponCode);
```

Comment:
I would suggest to do it above the message, to if something goes wrong with setting, it happens before we send out the message.

besides that I think it would make more sende, to directly set en empty string, or even Null(does Null work there?)
The empty string, because if the else is triggered, this means $codeLength did resolve to false